### DISCUSSION TITLE
How to create a JSON Web Token in C# .NET desktop

### DISCUSSION THREAD

**Discussion by harrymattison** [community/community#134870] [Programming Help]
### Body

Hi - I'm hoping that I can use a Github App to allow my C# desktop application to read from a private repo
 I've read this https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/generating-a-json-web-token-jwt-for-a-github-app and trying to use the sample code at https://docs.hidglobal.com/dev/auth-service/buildingapps/csharp/create-and-sign-a-json-web-token--jwt--with-c--and--net.htm

But Convert.FromBase64String throws "The input is not a valid Base-64 string as it contains a non-base 64 character, more than two padding characters, or an illegal character among the padding characters."
![image](https://github.com/user-attachments/assets/18f96448-5799-414b-bc2f-21d6909294b1)

Thanks for your help!

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by vanh01** [comment]
Maybe you replaced the wrong character.

![WXWorkCapture_17229984021189](https://github.com/user-attachments/assets/671054ad-b73b-4233-b8c3-7b04bfdb54d4)

Please pay more attention to the `\n` character

**Comment by ricaun** [comment]
This works better to create the jwt token: https://www.stevejgordon.co.uk/authenticating-a-github-app-using-a-json-web-token-in-dotnet

Or something like this:

```c#
/// <summary>
/// Generates a JWT token using the provided private key and client ID.
/// </summary>
/// <param name="private_key">The private key used for signing the token.</param>
/// <param name="client_id">The client ID associated with the token.</param>
/// <param name="minute_expires">The number of minutes until the token expires. Default is 10 minutes.</param>
/// <remarks><code>
/// string privateKey = File.ReadAllText("private-key.pem");
/// string clientId = "012345789";
/// string jwt = GenerateJwtToken(privateKey, clientId);
/// </code></remarks>
private string GenerateJwtToken(string private_key, string client_id, double minute_expires = 10)
{
    var rsa = RSA.Create();
    rsa.ImportFromPem(private_key);
    var signingCredentials = new SigningCredentials(new RsaSecurityKey(rsa), SecurityAlgorithms.RsaSha256);

    var jwtSecurityTokenHandler = new JwtSecurityTokenHandler { SetDefaultTimesOnTokenCreation = false };

    var now = DateTime.UtcNow.AddSeconds(-60);

    var jwt = jwtSecurityTokenHandler.CreateToken(new SecurityTokenDescriptor
    {
        Issuer = client_id,
        Expires = now.AddMinutes(minute_expires),
        IssuedAt = now,
        SigningCredentials = signingCredentials
    });

    return new JwtSecurityTokenHandler().WriteToken(jwt);
}
```
